vignettes/target_types.Rmd

```diff
@@ -0,0 +1,49 @@
+---
+title: "Target type guidelines"
+author: "David Watkins"
+date: "3/25/2019"
+output: rmarkdown::html_vignette
+vignette: >
+  %\VignetteIndexEntry{Target types}
+  %\VignetteEngine{knitr::rmarkdown}
+  \usepackage[utf8]{inputenc}
+---
+
+```{r setup, include=FALSE}
+knitr::opts_chunk$set(echo = TRUE)
+```
+
+`scipiper` targets can be indicator files with a [shared cache](shared_cache.html), local data files with or without local indicator files, or simply R objects.  This vignette provides some guidelines for when to use these different patterns.
+
+## Objectives
+There are two main goals to keep in mind when deciding what style of target to use:
+ 
+ - **Minimize time spend unnecessarily rebuilding targets** --- i.e., those that have already been built locally or by collaborators
+ - **Keeping the pipeline simple** --- caching can save lots of build time, but also adds to the code that has to be maintained, can clutter the repo with build and indicator files, etc.
+ 
+ These two can sometimes be in tension, but usually it will be clear which is more important for a given case.
+ 
+## Key factors to consider
+
+ - **Time to create an object** vs **object size**: This is the biggest thing to consider --- is it faster to upload a file to the shared cache and have others download it, or just have others rebuild it themselves?  This will somewhat depend on your internet connection.  Don't forget about upload time as well.  Uploading is generally slower than downloading, especially for remote workers with residential internet connections.   
+ 
+ - **How often will a target be updated?**:  Saving a target to disk may be less useful if it is likely to need to be rebuilt before the file on disk is ever used again.
+ 
+ - **Collaborators**: Are all collaborators equipped to rebuild every step in the pipeline, if needed?  If a particular step needs to be done by certain people, say a processing-intensive munging step early in the pipeline, than this file will almost certainly need to be in the remote cache.
+ 
+## Target choices
+
+**Shared cache data/indicator file**: The target will be only be rebuilt if its dependencies have changed, and the indicator file has not been updated to trigger a download from the remote cache. For longer-running operations that are worthwhile to cache remotely.  See the [shared caching vignette](shared_cache.html) for different implementation options.
+
+**Local data file only**:  The target will be rebuild if the local data file does not exist or depedencies have changed.  For cases where a target takes a relatively long time to build, but creates a massive file that is more cumbersome to upload/download than rebuild.  
+
+**Local data file with local indicator**: The target will be rebuilt if the indicator file is missing or dependencies have changed.  This is typically used to speed up `scipiper` by reducing redundant hashing of large files, or to represent multiple files, allowing a "one-to-many" target.
```

Comment:
This bullet may need elaboration for newcomers and posterity - can you make these examples more concrete and detailed?

Reply:
Do any of the recent features make the one-to-many application obsolete?

Reply:
Good Q. Actually, the one-to-many applications we've worked on in the last few months have used exactly this pattern, with one indicator file to represent many other files. Here's an example: https://github.com/USGS-R/lake-temperature-model-prep/blob/master/6_drivers/log/feather/6_fetch_new_cell_tasks.ind. This one is a complex example because it's an indicator file to represent indicator files for yml files, but a simpler example indicator file would look very similar, storing names and hashes of data files rather than other indicator files.